### PR TITLE
ENH: analytical entropy for invgauss distribution

### PR DESCRIPTION
**Pull request by dschmitz89**
#### Reference issue
#17748

#### What does this implement/fix?
An analytical expression for the entropy of the inverse gaussian distribution found [here](https://moser-isi.ethz.ch/docs/papers/smos-2012-10.pdf). As I understand, implementing $\mu$ is enough as the distribution infrastructure takes care of the scale (in the formula $\lambda$). For sake of review, the formula from the reference:

image

### CODE REVIEW

**Review by mdhaber — APPROVED** [review]
The paper does seem to be about the same inverse Gaussian distribution except their scale $\lambda$ doesn't appear to be the same as ours. In any case, that doesn't affect this PR because when we take $\lambda=1$ it's the same as our distribution with `loc = 0` and `scale = 1`. As you note, the distribution infrastructure takes care of the scale.

The formula here is not good for small `mu` (`mu < 0.008` or so) because `sc.exp1(2 / mu)` vanishes as `np.exp(2 / mu)` explodes. However, our existing implementation gives incorrect answers for `mu < 0.008`, too, so I'm not sure it's a huge loss. This is faster and probably more accurate for moderate `mu`, so it's probably fine until we implement a Padé approximant or something.  For large `mu`, it is fine up to about `mu=1e100` at which point `mu**3` will overflow. You could easily obviate that by breaking up the log of the product as a sum of two logs, turning `log(mu**3)` into `3*log(mu)`.

I think relying on the generic test for this is fine since it wasn't written specifically to avoid numerical problems.

One thing I'd ask you to do before I merge is to play around a bit with small `mu`. Make sure you agree that this is not much worse than what we have now for small `mu` and that there's not anything very easy we can do to fix it (besides some approximant). If you're still happy with it, I'll merge, and we can add an approximant for small `mu` in a follow-up.

**Comment by dschmitz89** [comment]
Thanks for the suggestions @mdhaber . Updated the formula to always logscale first. That gets rid of the problems for small $\mu$.

Plots comparing this PR and main:

InvGauss_Entropy_Small_mu
InvGauss_Entropy_Moderate_mu

**Comment by mdhaber** [comment]
Great. This is definitely better. I still think `exp1` will overflow before too long, but this undoubtedly improves the situation compared to `main`. 

Now the question is what kind of regressions we want to guard against. If we don't care whether an update to this code reverses this improvement for small and large `mu`, then the generic test is sufficient. If we do want to guard against that, then I think one test with small `mu` and one test with outrageously large `mu` would suffice. What do you think?

**Comment by WarrenWeckesser** [comment]
As noted above, this still fails when `mu` is small.  It appears to break down near `mu = 2e-3`:

```
In [1]: from scipy import stats

In [2]: mu = 2e-3

In [3]: stats.invgauss.entropy(mu)
/home/warren/repos/git/forks/scipy/build-install/lib/python3.10/site-packages/scipy/stats/_continuous_distns.py:4343: RuntimeWarning: divide by zero encountered in log
  b = np.exp(2 / mu + np.log(sc.exp1(2 / mu)))
Out[3]: -7.902973614428616

```
For comparison, I implemented the entropy function for the inverse Gaussian distribution in `mpsci`, and we can see that the above result gives only about 3 significant digits:
```
In [4]: from mpsci.distributions import invgauss

In [5]: from mpmath import mp

In [6]: mp.dps = 150

In [7]: float(invgauss.entropy(mu))
Out[7]: -7.904472117419651
```

Over in https://github.com/scipy/scipy/pull/17844, I implemented `scipy.special.logexp1` to compute $\log(E_1(x))$.  If we use that here, we can compute the entropy much more accurately:

```
In [1]: from scipy.special import logexp1

In [2]: def invgauss_entropy(mu):
   ...:     return (0.5 * (np.log(2 * np.pi) + 3*np.log(mu) + 1) -
   ...:             1.5 * np.exp(2/mu + logexp1(2/mu)))
   ...: 

In [3]: mu = 2e-3

In [4]: invgauss_entropy(mu)
Out[4]: -7.904472117419652
```

Whatever we do, the function definitely needs a test or two for small mu, where we know there are problems.

**Comment by dschmitz89** [comment]
I think this can be merged even without the new dedicated special function. Updating the entropy once more will be quite easy once it becomes available and this implementation is already an improvement.

File: scipy/stats/tests/test_distributions.py [hunk 1]
```diff
@@ -2470,6 +2470,11 @@ def test_logcdf_logsf(self):
         logsf = stats.invgauss.logsf(110, 1.05)
         assert_allclose(logsf, -56.1467092416426)
 
+    @pytest.mark.parametrize("mu, ref", [(1e-2, -5.496279615296405),
+                                         (1e8, 3.3244822568873467)])
+    def test_entropy(self, mu, ref):
+        assert_allclose(stats.invgauss.entropy(mu), ref)
```
[the remark this answers is not in the document]
Comment (mdhaber) [reply scipy/stats/tests/test_distributions.py:2476]:
```suggestion
    # from mpmath import mp
    # mp.dps = 100
    # mu = mp.mpf(1e-2)
    # ref = (1/2 * mp.log(2 * mp.pi * mp.e * mu**3)
    #        - 3/2* mp.exp(2/mu) * mp.e1(2/mu))
    @pytest.mark.parametrize("mu, ref", [(1e-2, -5.496279615262233),
                                         (1e8, 3.3244822568873474)])
                                         (1e100, 3.3244828013968899)])
    def test_entropy(self, mu, ref):
        assert_allclose(stats.invgauss.entropy(mu), ref)
```

File: scipy/stats/tests/test_distributions.py [hunk 2]
```diff
@@ -2470,6 +2470,17 @@ def test_logcdf_logsf(self):
...
         assert_allclose(logsf, -56.1467092416426)
 
+    # from mpmath import mp
+    # mp.dps = 100
+    # mu = mp.mpf(1e-2)
+    # ref = (1/2 * mp.log(2 * mp.pi * mp.e * mu**3)
+    #        - 3/2* mp.exp(2/mu) * mp.e1(2/mu))
+    @pytest.mark.parametrize("mu, ref", [(1e-2, -5.496279615262233),
+                                         (1e8, 3.3244822568873474)])
+                                         (1e100, 3.3244828013968899)])
```
Comment (mdhaber) [inline scipy/stats/tests/test_distributions.py:2480]: We get precisely the same results with mpmath numerical integration:
```python3
import numpy as np
from mpmath import mp
mp.dps = 100

def _logpdf(x, mu):
    # SciPy's definition of _logpdf
    return -0.5 * mp.log(2 * mp.pi) - 1.5 * mp.log(x) - ((x - mu) / mu) ** 2 / (2 * x)

def logpdf(x, mu, loc, scale):
    return _logpdf((x - loc)/scale, mu) - mp.log(scale)

@np.vectorize
def entropy(mu, loc=0.0, scale=1.0):
    mu, loc, scale = mp.mpf(mu), mp.mpf(loc), mp.mpf(scale)
    def integrand(x):
        logf = logpdf(x, mu, loc, scale)
        f = mp.exp(logf)
        return -f * logf
    return mp.quad(integrand, (loc, loc+mu, np.inf))

entropy(0.01)

# array(mpf('-5.496279615262233314139570946311748385591116312369781831539906340907631749051583759839841056064307524042'), dtype=object)
```
It agrees with the formula in all but the last two of the 103 digits.

File: scipy/stats/tests/test_distributions.py [hunk 3]
```diff
@@ -2470,6 +2470,17 @@ def test_logcdf_logsf(self):
         logsf = stats.invgauss.logsf(110, 1.05)
         assert_allclose(logsf, -56.1467092416426)
 
+    # from mpmath import mp
+    # mp.dps = 100
+    # mu = mp.mpf(1e-2)
+    # ref = (1/2 * mp.log(2 * mp.pi * mp.e * mu**3)
+    #        - 3/2* mp.exp(2/mu) * mp.e1(2/mu))
+    @pytest.mark.parametrize("mu, ref", [(1e-2, -5.496279615262233),
+                                         (1e8, 3.3244822568873474)])
```
Comment (mdhaber) [inline scipy/stats/tests/test_distributions.py:2479]:
```suggestion
                                         (1e8, 3.3244822568873474)]),
```

File: scipy/stats/tests/test_distributions.py [hunk 4]
```diff
@@ -2470,6 +2470,17 @@ def test_logcdf_logsf(self):
...
         assert_allclose(logsf, -56.1467092416426)
 
+    # from mpmath import mp
+    # mp.dps = 100
+    # mu = mp.mpf(1e-2)
+    # ref = (1/2 * mp.log(2 * mp.pi * mp.e * mu**3)
+    #        - 3/2* mp.exp(2/mu) * mp.e1(2/mu))
+    @pytest.mark.parametrize("mu, ref", [(1e-2, -5.496279615262233),
+                                         (1e8, 3.3244822568873474)]),
+                                         (1e100, 3.3244828013968899)])
```
Comment (mdhaber) [inline scipy/stats/tests/test_distributions.py:2480]:
```suggestion
    @pytest.mark.parametrize("mu, ref", [(1e-2, -5.496279615262233),
                                         (1e8, 3.3244822568873474)]),
                                         (1e100, 3.3244828013968899)])
```

File: scipy/stats/tests/test_distributions.py (hunk 2 above)
Comment (dschmitz89) [reply scipy/stats/tests/test_distributions.py:2480]: Nice trick with @np.vectorize! I actually computed those values using `mp.quad` too, just forgot to add that to the comments.  Sry that you had to implement it again.

File: scipy/stats/tests/test_distributions.py [hunk 5]
```diff
@@ -2470,6 +2470,17 @@ def test_logcdf_logsf(self):
         logsf = stats.invgauss.logsf(110, 1.05)
         assert_allclose(logsf, -56.1467092416426)
 
+    # from mpmath import mp
+    # mp.dps = 100
+    # mu = mp.mpf(1e-2)
+    # ref = (1/2 * mp.log(2 * mp.pi * mp.e * mu**3)
+    #        - 3/2* mp.exp(2/mu) * mp.e1(2/mu))
+    @pytest.mark.parametrize("mu, ref", [(1e-2, -5.496279615262233),
+                                         (1e8, 3.3244822568873474)]),
```
Comment (mdhaber) [inline scipy/stats/tests/test_distributions.py:2479]:
```suggestion
                                         (1e8, 3.3244822568873474),
```

**Comment by mdhaber** [comment]
Wow I really botched that. Sorry for all the noise.

File: scipy/stats/tests/test_distributions.py (hunk 2 above)
Comment (mdhaber) [reply scipy/stats/tests/test_distributions.py:2480]: No, I already had this. Actually, I'll open a PR to add an `mpmath` distribution interface to the top of `test_distributions.py` so we don't need to repeat the same sort of code everywhere.

**Comment by dschmitz89** [comment]
> Wow I really botched that. Sorry for all the noise.

No worries, programming often becomes way too quiet. ;)

**Comment by mdhaber** [comment]
Thanks @dschmitz89. Feel free to note in gh-17748 or gh-17844 that the second term can be reworked when gh-17844 merges. I don't think it needs to be tracked as a separate issue.

**Pull request merged**